File: C++/find-missing-observations.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    vector<int> missingRolls(vector<int>& rolls, int mean, int n) {
        static const int MAX_V = 6;
        static const int MIN_V = 1;
        
        const int total = accumulate(cbegin(rolls), cend(rolls), 0);
        const int missing = mean * (n + size(rolls)) - total;
        if (missing < MIN_V * n || missing > MAX_V * n) {
            return {};
        }
        const int q = missing / n;
        const int r = missing % n;
        vector<int> result(n);
        for (int i = 0; i < size(result); ++i) {
            result[i] = q + int(i < r);
        }
        return result;
    }
};


// Time:  O(n)
// Space: O(1)
class Solution2 {
public:
    vector<int> missingRolls(vector<int>& rolls, int mean, int n) {
        static const int MAX_V = 6;
        static const int MIN_V = 1;
        
        const int total = accumulate(cbegin(rolls), cend(rolls), 0);
        const int missing = mean * (n + size(rolls)) - total;
        if (missing < MIN_V * n || missing > MAX_V * n) {
            return {};
        }
        const int q = (missing - MIN_V * n) / (MAX_V - MIN_V);
        const int r = (missing - MIN_V * n) % (MAX_V - MIN_V);
        vector<int> result(n);
        for (int i = 0; i < size(result); ++i) {
            result[i] = (i < q) ? MAX_V : ((i == q) ? MIN_V + r : MIN_V);
        }
        return result;
    }
};

File: Python/find-missing-observations.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def missingRolls(self, rolls, mean, n):
        """
        :type rolls: List[int]
        :type mean: int
        :type n: int
        :rtype: List[int]
        """
        MAX_V = 6
        MIN_V = 1
        total = sum(rolls)
        missing = mean*(n+len(rolls))-total
        if missing < MIN_V*n or missing > MAX_V*n:
            return []
        q, r = divmod(missing, n)
        return [q+int(i < r) for i in xrange(n)]


# Time:  O(n)
# Space: O(1)
class Solution2(object):
    def missingRolls(self, rolls, mean, n):
        """
        :type rolls: List[int]
        :type mean: int
        :type n: int
        :rtype: List[int]
        """
        MAX_V = 6
        MIN_V = 1
        total = sum(rolls)
        missing = mean*(n+len(rolls))-total
        if missing < MIN_V*n or missing > MAX_V*n:
            return []
        q, r = divmod(missing-MIN_V*n, (MAX_V-MIN_V))
        return [MAX_V if i < q else MIN_V+r if i == q else MIN_V  for i in xrange(n)]
